### DISCUSSION TITLE
Clear pins from map

### DISCUSSION THREAD

**Discussion by stevegdavis** [Mapsui/Mapsui#1762] [Q&A] [answered]
I have created a pin on the map (Maui app) but I want to update the pin details such as lat,lon label text etc... every so often from a server.
I have tried to clear all the pins from the map view eg. mapView.Pins.Clear(); but all the previous pin(s) are still on the map despite the pins list being empty. Code snippet:-

var list = await webHelper.DevicesFetchAsync(parameters);
mapView.Pins.Clear();
            foreach (var device in list.Devices)
            {
                var myPin = new Pin(mapView)
                {
                    Position = new Mapsui.UI.Maui.Position(device.Latitude, device.Longitude),
                    Type = PinType.Pin,
                    Label = $"{device.Name}\n{device.Date}",
                    Address = "",
                    Scale = 0.643F,
                    Color = Color.FromArgb("#248f8f")
                };
                myPin.ShowCallout();
                myPin.Callout.Title = $"{device.Name}\n{device.Date}";
                myPin.Callout.ArrowHeight = 10;
                myPin.Callout.TitleFontSize = 15;
                //myPin.Callout.Color = c;
                //myPin.Callout.TitleFontColor = c;
                mapView.Pins.Add(myPin);
            }
            mapView.Refresh(ChangeType.Continuous);
            mapView.Refresh();
            mapView.RefreshGraphics();

Screenshot-
![Screenshot_20221118-151217](https://user-images.githubusercontent.com/15140454/202749457-eed2ce30-4564-4089-a0e6-ac1be795c8a9.jpg)

If I click on the pins they disappear, until only the latest remains on the map.
If I just call mapView.Pins.Clear() and not replace any pins then the pins on the map do disappear.
Thanks.

**Comment by charlenni** [comment]
Do you get any error for the `Clear()` command? Use the logging for this.

**Reply by stevegdavis** [reply]
Many thanks for your reply, do you mean any exceptions as the Clear() is generic list clear, I would not assume any error at that point, is that what you mean? any pointers for logging appreciated.

**Comment by charlenni** [comment]
You could find information about logging [here](http://mapsui.com/documentation/logging.html).

Behind the scenes, the `Clear()` creates an event, which updates things on the screen. There could errors be created.

**Reply by stevegdavis** [reply]
No I cannot see any logged errors etc... related to the clear just 'The CancellationTokenSource has been disposed.' not sure where that comes from and a warning 'An error occurred while fetching data' again not sure where that comes from either and only some info messages relating to the time to draw a control.

**Reply by inforithmics** [reply]
Which version of Mapsui are you using?

**Reply by stevegdavis** [reply]
4.0.0-beta.5 and .NET7

**Comment by inforithmics** [comment]
I discovered in another Control that .NET 6 Maui and .NET 7 Maui is not binary compatible. This means that Maui Controls need to be compiled against .NET 6 Maui and .NET 7 Maui. 

See here:
https://github.com/Mapsui/Mapsui/pull/1749

**Reply by inforithmics** [reply]
Can you try this (nugets compiled against main) https://ci.appveyor.com/project/pauldendulk/mapsui/builds/45436440/artifacts

**Reply by stevegdavis** [reply]
Sorry tried the package but I get a load of errors, I noticed that there are some dependencies missing such as SkiaSharp.Views.Maui.Controls
from the package, installed those (latest version) that cured those errors but still some others, screenshot if it helps:-



![Screenshot 2022-11-22 100850](https://user-images.githubusercontent.com/15140454/203289088-af44cea7-c1dd-4071-8555-ebd5188226e1.png)

**Reply by inforithmics** [reply]
When you used the new nuget package have you cleaned the obj Folder (because the nuget dependencies are cached in this folder)

**Reply by inforithmics** [reply]
I've seen you are using a Test Project can you attach it in a comment?

**Reply by stevegdavis** [reply]
Yes did the usual bin/obj delete to no avail, project zip attached.
[Test_Mapsui.zip](https://github.com/Mapsui/Mapsui/files/10069455/Test_Mapsui.zip)

**Reply by inforithmics** [reply]
1. You forgot to reference Mapsui.Maui that's the errors.
2. You have to add a nuget.config with
```
<?xml version="1.0" encoding="utf-8"?>

<configuration>
	<!-- Define the package sources, nuget.org and contoso.com. -->
	<!-- `clear` ensures no additional sources are inherited from another config file. -->
	<packageSources>
	  <clear />
	  <!-- `key` can be any identifier for your source. -->
	  <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />  
      <add key="LocalNuget" value="./packages/" />
	</packageSources>	
</configuration>
```

localnuget being the path where the downloaded nuget packages are.

3. I looked at the samples and with the newest version it seems to work.

4. Here is the fixed sample project.
[Fixed_Test_Mapsui.zip](https://github.com/Mapsui/Mapsui/files/10071664/Fixed_Test_Mapsui.zip)

**Comment by stevegdavis** [comment, marked as answer]
@inforithmics Many apologies - if I set
foreach (var pin in mapView.Pins)
        {
            pin.HideCallout();
        }
before I clear the mapview pins it all works as I require, even with 4.0.0-beta.5.
Again many apologies I hope I didn't waste too much of your time.

**Reply by inforithmics** [reply]
No problem was a good test if the .Net 7 maui target would work. I'll have a look int the pin.hidecallout logic.

**Reply by stevegdavis** [reply]
OK, good luck.
